python3Packages.pandas: tests fail on master

**Describe the bug**
Build fails with the following single test failure:

```
=================================== FAILURES ===================================
__________________________ TestAccessor.test_from_coo __________________________

self = <pandas.tests.arrays.sparse.test_array.TestAccessor object at 0x7fff6d82a3a0>

    @td.skip_if_no_scipy
    def test_from_coo(self):
        import scipy.sparse

        row = [0, 3, 1, 0]
        col = [0, 3, 1, 2]
        data = [4, 5, 7, 9]
        sp_array = scipy.sparse.coo_matrix((data, (row, col)))
        result = pd.Series.sparse.from_coo(sp_array)

        index = pd.MultiIndex.from_arrays([[0, 0, 1, 3], [0, 2, 1, 3]])
        expected = pd.Series([4, 9, 7, 5], index=index, dtype="Sparse[int]")
>       tm.assert_series_equal(result, expected)
E       AssertionError: Attributes of Series are different
E       
E       Attribute "dtype" are different
E       [left]:  Sparse[float64, nan]
E       [right]: Sparse[int64, 0]

tests/arrays/sparse/test_array.py:1196: AssertionError
```

**Notify maintainers**
<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

@7c6f434c @fridh @knedlsepp

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: python3Packages.pandas
# a list of nixos modules affected by the problem
module:
```

--- a/pkgs/development/python-modules/pandas/default.nix
+++ b/pkgs/development/python-modules/pandas/default.nix
@@ -34,6 +34,8 @@ buildPythonPackage rec {
     inherit pname version;
     sha256 = "06vhk75hmzgv1sfbjzgnsw9x10h7y6bd6s6z7d6lfnn7wcgc83zi";
   };
+  # See https://github.com/scipy/scipy/issues/13585 and https://github.com/pandas-dev/pandas/pull/40020
+  patches = [ ./fix-tests.patch ];
 
   nativeBuildInputs = [ cython ];
   buildInputs = lib.optional stdenv.isDarwin libcxx;
@@ -58,16 +60,6 @@ buildPythonPackage rec {
   # doesn't work with -Werror,-Wunused-command-line-argument
   # https://github.com/NixOS/nixpkgs/issues/39687
   hardeningDisable = lib.optional stdenv.cc.isClang "strictoverflow";
-
-  # For OSX, we need to add a dependency on libcxx, which provides
-  # `complex.h` and other libraries that pandas depends on to build.
-  postPatch = lib.optionalString stdenv.isDarwin ''
-    cpp_sdk="${libcxx}/include/c++/v1";
-    echo "Adding $cpp_sdk to the setup.py common_include variable"
-    substituteInPlace setup.py \
-      --replace "['pandas/src/klib', 'pandas/src']" \
-                "['pandas/src/klib', 'pandas/src', '$cpp_sdk']"
-  '';
 
   # Parallel Cythonization is broken in Python 3.8 on Darwin. Fixed in the next
   # release. https://github.com/pandas-dev/pandas/pull/30862

--- a/pkgs/development/python-modules/pandas/fix-tests.patch
+++ b/pkgs/development/python-modules/pandas/fix-tests.patch
@@ -1,0 +1,12 @@
+diff --color -ur a/pandas/tests/arrays/sparse/test_array.py b/pandas/tests/arrays/sparse/test_array.py
+--- a/pandas/tests/arrays/sparse/test_array.py	2020-12-07 12:42:08.000000000 +0100
++++ b/pandas/tests/arrays/sparse/test_array.py	2021-02-27 21:48:16.483903149 +0100
+@@ -1188,7 +1188,7 @@
+         row = [0, 3, 1, 0]
+         col = [0, 3, 1, 2]
+         data = [4, 5, 7, 9]
+-        sp_array = scipy.sparse.coo_matrix((data, (row, col)))
++        sp_array = scipy.sparse.coo_matrix((data, (row, col)), dtype="int")
+         result = pd.Series.sparse.from_coo(sp_array)
+ 
+         index = pd.MultiIndex.from_arrays([[0, 0, 1, 3], [0, 2, 1, 3]])

